src/parquet/file_reader.cc

```diff
@@ -0,0 +1,764 @@
+            if constexpr (std::is_trivially_copyable_v<T>) {
+                out[completed + i] = _dict[buf[i]];
+            } else {
+                // seastar::temporary_buffer<uint8_t>
+                out[completed + i] = _dict[buf[i]].share();
+            }
+        }
+        completed += n_read;
+        if (n_read < n_to_read) {
+            return completed;
+        }
+    }
+    return n;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// rle_decoder_boolean
+
+void rle_decoder_boolean::reset(std::basic_string_view<uint8_t> data) {
+    _rle_decoder.Reset(data.data(), data.size(), 1);
+}
+
+size_t rle_decoder_boolean::read_batch(size_t n, uint8_t out[]) {
+    return _rle_decoder.GetBatch(out, n);
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+// delta_binary_packed_decoder
+
+template <typename T>
+void delta_binary_packed_decoder<T>::reset(std::basic_string_view<uint8_t> data) {
+    _decoder.Reset(data.data(), data.size());
+    _values_current_block = 0;
+    _values_current_mini_block = 0;
+}
+
+template <typename T>
+void delta_binary_packed_decoder<T>::init_block() {
+    int32_t block_size;
+    if (!_decoder.GetVlqInt(&block_size)) throw parquet_exception::eof();
+    if (!_decoder.GetVlqInt(&_num_mini_blocks)) throw parquet_exception::eof();
+    if (!_decoder.GetVlqInt(&_values_current_block)) throw parquet_exception::eof();
+    if (!_decoder.GetZigZagVlqInt(&_last_value)) throw parquet_exception::eof();
+
+    if (_delta_bit_widths.size() < static_cast<size_t>(_num_mini_blocks)) {
+        _delta_bit_widths = buffer(_num_mini_blocks);
+    }
+
+    if (!_decoder.GetZigZagVlqInt(&_min_delta)) throw parquet_exception::eof();
+    for (int i = 0; i < _num_mini_blocks; ++i) {
+        if (!_decoder.GetAligned<uint8_t>(1, _delta_bit_widths.data() + i)) {
+            throw parquet_exception::eof();
+        }
+    }
+    _values_per_mini_block = block_size / _num_mini_blocks;
+    _mini_block_idx = 0;
+    _delta_bit_width = _delta_bit_widths.data()[0];
+    _values_current_mini_block = _values_per_mini_block;
+}
+
+template <typename T>
+size_t delta_binary_packed_decoder<T>::read_batch(size_t n, T out[]) {
+    for (size_t i = 0; i < n; ++i) {
+        if (__builtin_expect(_values_current_mini_block == 0, 0)) {
+            ++_mini_block_idx;
+            if (_mini_block_idx < static_cast<size_t>(_num_mini_blocks)) {
+                _delta_bit_width = _delta_bit_widths.data()[_mini_block_idx];
+                _values_current_mini_block = _values_per_mini_block;
+            } else {
+                init_block();
+                out[i] = _last_value;
+                continue;
+            }
+        }
+
+        // TODO: the key to this algorithm is to decode the entire miniblock at once.
+        int64_t delta;
+        if (!_decoder.GetValue(_delta_bit_width, &delta)) throw parquet_exception::eof();
```

Comment:
please wrap all `if` statements in curly braces, that's our strict code style

Reply:
Fixed in ac9cdae.